frontend/app/components/context-menus/types-context-menu/types-context-menu.servcie.ts

```diff
@@ -0,0 +1,60 @@
+//-- copyright
+// OpenProject is a project management system.
+// Copyright (C) 2012-2015 the OpenProject Foundation (OPF)
+//
+// This program is free software; you can redistribute it and/or
+// modify it under the terms of the GNU General Public License version 3.
+//
+// OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+// Copyright (C) 2006-2013 Jean-Philippe Lang
+// Copyright (C) 2010-2013 the ChiliProject Team
+//
+// This program is free software; you can redistribute it and/or
+// modify it under the terms of the GNU General Public License
+// as published by the Free Software Foundation; either version 2
+// of the License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program; if not, write to the Free Software
+// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+//
+// See doc/COPYRIGHT.rdoc for more details.
+//++
+
+import {opWorkPackagesModule} from '../../../angular-modules';
+import {CollectionResource} from '../../api/api-v3/hal-resources/collection-resource.service';
+import {HalResource} from '../../api/api-v3/hal-resources/hal-resource.service';
+
+class TypesContextMenuController {
+  public types = [];
+
+  constructor(protected $state, protected $scope, v3Path, halRequest) {
+    const project = $scope.projectIdentifier;
+    const typesUrl = v3Path.types({project});
+    $scope.$ctrl = this;
+
+    halRequest.get(typesUrl).then(types => this.types = types.elements);
```

Comment:
This gets the types from the `api/v3/types` endpoint which will return all types regardless of whether they are activated in the project or not. When deactivating a type in a project, it is still available in the list of types the user can choose from. If the user clicks on a type which is not available, the first type in the list is selected instead. 

In order to get the types that are active in the project, one would have to send a POST to `/api/v3/projects/:project_id/work_packages/form` and retrieve the types from there.

This is unfortunately not desirable when the project context is missing, e.g. in the global work packages list. We could either:
* remove the create button there altogether
* send a post to `api/v3/work_packages/form` instead. For this we would have to send a project with the request which requires a second request to `/api/v3/work_packages/available_projects` preceding the one to the form. As stated, this would require two requests and would give us only the types available in the first project.
* add an api endpoint analogue to `/api/v3/work_packages/available_projects` titled e.g. `/api/v3/work_packages/available_types` which returns all the types active in any project the user has access to.  To be concise, we would also add an `/api/v3/work_packages/:id/available_types` and `/api/v3/projects/:project_id/work_packages/available_types`. Doing that we could always call the `available_types` endpoint regardless of whether we are in a project or not.

Personally, I'd opt for adding an endpoint.

Apart from that, caused by the delay inevitable when making a backend request, the drop-down list is displayed as empty until the request returns and angular updates the list. A spinner would help to avoid the user wondering what happened to his click until the update takes place.